File: frontend/src/services/followService.ts
import { toast } from 'sonner';

const API_BASE_URL = 'http://localhost:8080/api';

export interface FollowResponse {
  success: boolean;
  message: string;
  follow?: {
    id: string;
    follower: { id: string; name: string };
    following: { id: string; name: string };
    createdAt: string;
  };
}

export interface IsFollowingResponse {
  isFollowing: boolean;
}

export interface FollowerCountResponse {
  count: number;
}

export interface FollowingCountResponse {
  count: number;
}

export interface FollowDTO {
  id: string;
  follower: { id: string; name: string; avatar?: string };
  following: { id: string; name: string; avatar?: string };
  isMuted: boolean;
  createdAt: string;
}

export interface PaginatedResponse<T> {
  content: T[];
  totalElements: number;
  totalPages: number;
  currentPage: number;
}

export const followService = {
  /**
   * Follow a user
   */
  async followUser(followerId: string, followingId: string): Promise<FollowResponse> {
    try {
      const response = await fetch(
        `${API_BASE_URL}/follows/${followerId}/follow/${followingId}`,
        {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
          },
        }
      );

      const data = await response.json();
      if (!response.ok) {
        throw new Error(data.message || 'Failed to follow user');
      }
      return data;
    } catch (error) {
      console.error('Error following user:', error);
      throw error;
    }
  },

  /**
   * Unfollow a user
   */
  async unfollowUser(followerId: string, followingId: string): Promise<FollowResponse> {
    try {
      const response = await fetch(
        `${API_BASE_URL}/follows/${followerId}/unfollow/${followingId}`,
        {
          method: 'DELETE',
          headers: {
            'Content-Type': 'application/json',
          },
        }
      );

      const data = await response.json();
      if (!response.ok) {
        throw new Error(data.message || 'Failed to unfollow user');
      }
      return data;
    } catch (error) {
      console.error('Error unfollowing user:', error);
      throw error;
    }
  },

  /**
   * Check if a user is following another user
   */
  async isFollowing(followerId: string, followingId: string): Promise<boolean> {
    try {
      const response = await fetch(
        `${API_BASE_URL}/follows/${followerId}/is-following/${followingId}`
      );

      if (!response.ok) {
        throw new Error('Failed to check follow status');
      }

      const data = (await response.json()) as IsFollowingResponse;
      return data.isFollowing;
    } catch (error) {
      console.error('Error checking follow status:', error);
      return false;
    }
  },

  /**
   * Get follower count for a user
   */
  async getFollowerCount(userId: string): Promise<number> {
    try {
      const response = await fetch(`${API_BASE_URL}/follows/${userId}/follower-count`);

      if (!response.ok) {
        throw new Error('Failed to get follower count');
      }

      const data = (await response.json()) as FollowerCountResponse;
      return data.count;
    } catch (error) {
      console.error('Error getting follower count:', error);
      return 0;
    }
  },

  /**
   * Get following count for a user
   */
  async getFollowingCount(userId: string): Promise<number> {
    try {
      const response = await fetch(`${API_BASE_URL}/follows/${userId}/following-count`);

      if (!response.ok) {
        throw new Error('Failed to get following count');
      }

      const data = (await response.json()) as FollowingCountResponse;
      return data.count;
    } catch (error) {
      console.error('Error getting following count:', error);
      return 0;
    }
  },

  /**
   * Get followers of a user
   */
  async getFollowers(
    userId: string,
    page: number = 0,
    size: number = 20
  ): Promise<PaginatedResponse<FollowDTO>> {
    try {
      const response = await fetch(
        `${API_BASE_URL}/follows/${userId}/followers?page=${page}&size=${size}`
      );

      if (!response.ok) {
        throw new Error('Failed to get followers');
      }

      return await response.json();
    } catch (error) {
      console.error('Error getting followers:', error);
      throw error;
    }
  },

  /**
   * Get following list for a user
   */
  async getFollowing(
    userId: string,
    page: number = 0,
    size: number = 20
  ): Promise<PaginatedResponse<FollowDTO>> {
    try {
      const response = await fetch(
        `${API_BASE_URL}/follows/${userId}/following?page=${page}&size=${size}`
      );

      if (!response.ok) {
        throw new Error('Failed to get following list');
      }

      return await response.json();
    } catch (error) {
      console.error('Error getting following list:', error);
      throw error;
    }
  },

  /**
   * Mute a user
   */
  async muteUser(followerId: string, followingId: string): Promise<FollowResponse> {
    try {
      const response = await fetch(
        `${API_BASE_URL}/follows/${followerId}/mute/${followingId}`,
        {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
          },
        }
      );

      const data = await response.json();
      if (!response.ok) {
        throw new Error(data.message || 'Failed to mute user');
      }
      return data;
    } catch (error) {
      console.error('Error muting user:', error);
      throw error;
    }
  },

  /**
   * Unmute a user
   */
  async unmuteUser(followerId: string, followingId: string): Promise<FollowResponse> {
    try {
      const response = await fetch(
        `${API_BASE_URL}/follows/${followerId}/unmute/${followingId}`,
        {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
          },
        }
      );

      const data = await response.json();
      if (!response.ok) {
        throw new Error(data.message || 'Failed to unmute user');
      }
      return data;
    } catch (error) {
      console.error('Error unmuting user:', error);
      throw error;
    }
  },

  /**
   * Get muted follows for a user
   */
  async getMutedFollows(userId: string): Promise<FollowDTO[]> {
    try {
      const response = await fetch(`${API_BASE_URL}/follows/${userId}/muted`);

      if (!response.ok) {
        throw new Error('Failed to get muted follows');
      }

      return await response.json();
    } catch (error) {
      console.error('Error getting muted follows:', error);
      return [];
    }
  },
};

File: backend/src/main/java/com/lifeflow/backend/services/FollowService.java
package com.lifeflow.backend.services;

import com.lifeflow.backend.dto.FollowDTO;
import com.lifeflow.backend.model.Follow;
import com.lifeflow.backend.model.User;
import com.lifeflow.backend.repository.FollowRepository;
import com.lifeflow.backend.repository.UserRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

@Service
@Transactional
public class FollowService {
    
    private final FollowRepository followRepository;
    private final UserRepository userRepository;
    private final NotificationService notificationService;
    
    public FollowService(FollowRepository followRepository, UserRepository userRepository, NotificationService notificationService) {
        this.followRepository = followRepository;
        this.userRepository = userRepository;
        this.notificationService = notificationService;
    }
    
    /**
     * Follow a user
     */
    public Follow followUser(String followerId, String followingId) {
        Optional<User> followerOpt = userRepository.findById(followerId);
        Optional<User> followingOpt = userRepository.findById(followingId);
        
        if (followerOpt.isEmpty() || followingOpt.isEmpty()) {
            throw new IllegalArgumentException("User not found");
        }
        
        if (followerId.equals(followingId)) {
            throw new IllegalArgumentException("Cannot follow yourself");
        }
        
        User follower = followerOpt.get();
        User following = followingOpt.get();
        
        // Check if already following
        if (followRepository.existsByFollowerAndFollowing(follower, following)) {
            throw new IllegalArgumentException("Already following this user");
        }
        
        Follow follow = Follow.builder()
                .follower(follower)
                .following(following)
                .isMuted(false)
                .build();
        
        Follow saved = followRepository.save(follow);
        
        // Trigger notification for new follower
        notificationService.createFollowerNotification(follower, following);
        
        return saved;
    }
    
    /**
     * Unfollow a user
     */
    public void unfollowUser(String followerId, String followingId) {
        Optional<User> followerOpt = userRepository.findById(followerId);
        Optional<User> followingOpt = userRepository.findById(followingId);
        
        if (followerOpt.isEmpty() || followingOpt.isEmpty()) {
            throw new IllegalArgumentException("User not found");
        }
        
        User follower = followerOpt.get();
        User following = followingOpt.get();
        
        Optional<Follow> followOpt = followRepository.findByFollowerAndFollowing(follower, following);
        if (followOpt.isPresent()) {
            followRepository.delete(followOpt.get());
        }
    }
    
    /**
     * Mute a user
     */
    public Follow muteUser(String followerId, String followingId) {
        Optional<User> followerOpt = userRepository.findById(followerId);
        Optional<User> followingOpt = userRepository.findById(followingId);
        
        if (followerOpt.isEmpty() || followingOpt.isEmpty()) {
            throw new IllegalArgumentException("User not found");
        }
        
        User follower = followerOpt.get();
        User following = followingOpt.get();
        
        Optional<Follow> followOpt = followRepository.findByFollowerAndFollowing(follower, following);
        if (followOpt.isEmpty()) {
            throw new IllegalArgumentException("Not following this user");
        }
        
        Follow follow = followOpt.get();
        follow.setIsMuted(true);
        follow.setMutedAt(LocalDateTime.now());
        
        return followRepository.save(follow);
    }
    
    /**
     * Unmute a user
     */
    public Follow unmuteUser(String followerId, String followingId) {
        Optional<User> followerOpt = userRepository.findById(followerId);
        Optional<User> followingOpt = userRepository.findById(followingId);
        
        if (followerOpt.isEmpty() || followingOpt.isEmpty()) {
            throw new IllegalArgumentException("User not found");
        }
        
        User follower = followerOpt.get();
        User following = followingOpt.get();
        
        Optional<Follow> followOpt = followRepository.findByFollowerAndFollowing(follower, following);
        if (followOpt.isEmpty()) {
            throw new IllegalArgumentException("Not following this user");
        }
        
        Follow follow = followOpt.get();
        follow.setIsMuted(false);
        follow.setMutedAt(null);
        
        return followRepository.save(follow);
    }
    
    /**
     * Check if user is following another user
     */
    public boolean isFollowing(String followerId, String followingId) {
        Optional<User> followerOpt = userRepository.findById(followerId);
        Optional<User> followingOpt = userRepository.findById(followingId);
        
        if (followerOpt.isEmpty() || followingOpt.isEmpty()) {
            return false;
        }
        
        return followRepository.existsByFollowerAndFollowing(followerOpt.get(), followingOpt.get());
    }
    
    /**
     * Get all followers of a user
     */
    public Page<FollowDTO> getFollowers(String userId, Pageable pageable) {
        Optional<User> userOpt = userRepository.findById(userId);
        if (userOpt.isEmpty()) {
            throw new IllegalArgumentException("User not found");
        }
        
        User user = userOpt.get();
        Page<Follow> followers = followRepository.findByFollowing(user, pageable);
        
        return followers.map(this::convertToDTO);
    }
    
    /**
     * Get all users that a user is following
     */
    public Page<FollowDTO> getFollowing(String userId, Pageable pageable) {
        Optional<User> userOpt = userRepository.findById(userId);
        if (userOpt.isEmpty()) {
            throw new IllegalArgumentException("User not found");
        }
        
        User user = userOpt.get();
        Page<Follow> following = followRepository.findByFollower(user, pageable);
        
        return following.map(this::convertToDTO);
    }
    
    /**
     * Get follower count
     */
    public long getFollowerCount(String userId) {
        Optional<User> userOpt = userRepository.findById(userId);
        if (userOpt.isEmpty()) {
            throw new IllegalArgumentException("User not found");
        }
        
        return followRepository.countByFollowing(userOpt.get());
    }
    
    /**
     * Get following count
     */
    public long getFollowingCount(String userId) {
        Optional<User> userOpt = userRepository.findById(userId);
        if (userOpt.isEmpty()) {
            throw new IllegalArgumentException("User not found");
        }
        
        return followRepository.countByFollower(userOpt.get());
    }
    
    /**
     * Get muted follows
     */
    public List<Follow> getMutedFollows(String userId) {
        Optional<User> userOpt = userRepository.findById(userId);
        if (userOpt.isEmpty()) {
            throw new IllegalArgumentException("User not found");
        }
        
        return followRepository.findMutedFollows(userOpt.get());
    }
    
    /**
     * Get active following users (non-muted)
     */
    public List<User> getActiveFollowingUsers(String userId) {
        Optional<User> userOpt = userRepository.findById(userId);
        if (userOpt.isEmpty()) {
            throw new IllegalArgumentException("User not found");
        }
        
        return followRepository.findActiveFollowingUsers(userOpt.get());
    }
    
    /**
     * Convert Follow entity to DTO
     */
    private FollowDTO convertToDTO(Follow follow) {
        User user = follow.getFollower().getId().equals(follow.getId()) ? follow.getFollowing() : follow.getFollower();
        
        return FollowDTO.builder()
                .id(follow.getId())
                .userId(user.getId())
                .userName(user.getName())
                .userEmail(user.getEmail())
                .userAvatar(user.getAvatar())
                .userBio(user.getBio())
                .isMuted(follow.getIsMuted())
                .createdAt(follow.getCreatedAt().toString())
                .build();
    }
}
